pkg/logs/internal/util/containersorpods/containers_or_pods.go

```diff
@@ -0,0 +1,193 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package containersorpods
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+)
+
+// LogWhat is the answer this package provides
+type LogWhat int
+
+const (
+	// LogContainers means that the logs-agent should log containers, not pods.
+	LogContainers LogWhat = iota
+
+	// LogPods means that the logs-agent should log pods, not containers.
+	LogPods
+
+	// LogUnknown indicates that it's too early to tell which should be used, because
+	// neither service has become available yet.
+	LogUnknown
+
+	// LogNothing means neither containers nor pods are supported.
+	LogNothing
+)
+
+type chooser struct {
+	// choice carries the result, once it is known.
+	choice chan LogWhat
+
+	// m covers the 'started' field
+	m sync.Mutex
+
+	// if started is true, then the process of choosing has begun
+	// and it is safe to wait for a value in `choice`.
+	started bool
+
+	// kubeletReady determines if kubelet is ready, or how long to wait
+	kubeletReady func() (bool, time.Duration)
+
+	// dockerReady determines if dockerutil is ready, or how long to wait
+	dockerReady func() (bool, time.Duration)
+}
+
+var singleton = chooser{
```

Comment:
I could have sworn I had a `SetChoice` method, but I can't find it!  It has been three weeks so memory has fades.

Your suggestion is a good one, and since Agent instantiates the Launchers directly, it's not too much work to thread the "Chooser" through.